### DISCUSSION TITLE
A good API for reactive props?

### DISCUSSION THREAD

**Discussion by calebporzio** [livewire/livewire#5612] [8. V3 (maintainers only to post)]
In the beginning... if you passed data into a component, the only way to access it would be through the `mount()` method:

```php
<livewire:show-post :id="$post->id" />
```

```php
class ShowPost extends Component
{
    public $id;

    public function mount($id)
    {
        $this->id = $id;
    }
}
```

When V2 launched, to cut down on boilerplate we made it so that "props" were automatically assigned to component public properties of the same name. So now it's:

```php
<livewire:show-post :id="$post->id" />
```

```php
class ShowPost extends Component
{
    public $id;
}
```

And `$id` will automatically be set on `ShowPost`.

And just so we have all the information we need to help solve the problem, pure Blade components don't do this "auto-setting", instead they pass props through a constructor like so:

```php
<livewire:show-post :id="$post->id" />
```

```php
class ShowPost extends Component
{
    public function __construct(
        public $id,
    ) {}
}
```

# The goal for V3
In V3, in addition to simply passing data into a component when it's created, there is also a concept of "reactive props", where certain props are passed in but marked as "reactive" and will automatically change when the parent data changes. This is similar to props in a framework like Vue or React.

Unlike Vue or React, making props reactive in Livewire adds some performance overhead. Every network request from the parent component must include the child component's snapshot, which makes the entire request slower.

So knowing this, it makes sense to make prop "reactivity" opt-in.

Here was my initial shot at an API:

```php
<livewire:show-post :id="$post->id" />
```

```php
class ShowPost extends Component
{
    #[Prop]
    public $id;
}
```

By adding the `#[Prop]` attribute, we tell Livewire to keep this property up-to-date if the parent value being passed in changes.

# The Problem
In V3, the word "prop" means two things now, so my intial API I feel is missleading.

A) Data passed in from the parent to be used in `mount()`
B) Data marked as "reactive" that will have performance implications

Let's talk solutions, there are many...

# Solution A) Stick with it, who cares

Maybe what I landed on by instinct is actually the best?

Pros: 
* It was intuitive to me, so it may be to others
* I demoed it at Laracon, so people are already a little "primed"
* The mental overhead of tracking what's "passed-in data" and what's a "(reactive) prop" isn't too bad?

Cons:
* Well, I already said earlier...

# B) Use the word "Reactive"
Instead of using the word `#[Prop]` for the attribute, use something like `#[Reactive]`

```php
class ShowPost extends Component
{
    #[Reactive]
    public $id;
}
```

Pro:
* Because we're not using "prop", this concept won't get convoluted with other props

Con:
* "Reactive" might be confusing as all properties are actually "reactive" in the sense that you can change them and Livewire "reacts"

# C) Use the word "Prop" AND "Reactive"

To address the con of "A", we could do something like:

```php
#[Prop, Reactive]
public $id;
```

Basically, we would force all "props" to be denoted as `#[Prop]`, and any ones we want reactive need the additional `#[Reactive]`

If we dislike the word "Reactive", we could find another word like: "Bound"

# D) Use a constructor
The reason we automatically assign properties in Livewire without the need for the `mount()` method is to cut down on unnescessary boilerplate. Blade components solved for this by using what PHP offers to cut down on the same boilerplate: That new constructor property syntax.

So in V3, we could use `__construct` instead of `mount` for receiving passed in data like:

```php
class ShowPost extends Component
{
    public __construct(
        public $id,
    ) {}
}
```

With this, we could remove the behavior of auto-assigning properties and reserve the word `#[Prop]` for "reactive props":

```php
class ShowPost extends Component
{
    public __construct(
        #[Prop] public $id,
        public $otherNonReactiveProperty,
    ) {}
}
```

Now, `$id` will be auto-assigned and reactive, and `$otherNonReactiveProperty` will be a standard property assigned through the constructor.

Pros:
* Can use `#[Prop]` with a clean conscience
* `__construct` is symmetrical with Blade components and more intuitive in general than `mount()`

Cons:
* `__construct` would need to be hacked to only run on the first request (like `mount`). You can instantiate a class without calling the constructor using Reflection, but it does feel weird to do something like that.
* It feels like there's just lots of other cons I'm not even thinking about lol.

# E) All props are reactive by default
This is almost definitely not going to happen but it's worth at least exploring.

We can make all properties passed in "reaactive". It would mean, across the board, every time a component was updated, all of it's children, and parents that properties are passed between would travel together from server to client.

This reduces the performance usefulness of nesting components, but increases their usefulness as an abstraction.

There could even be ways to make it a little smarter, like only "bind" two components together if the prop passed between them is a PHP dynamic prop, and not a static one. For example:

```php
<livewire:show-post id="1" :id="$post->id" />
```

In this example "id" is static, and ":id" is dynamic. "id" will never change, so we can safely not "bind" the two components in a request.

Also, we could add an attribute to opt-out of reactivity like:

```php
class ShowPost extends Component
{
    #[Stale]
    public $id;
}
```

I can't really think of a good name. `Static` feels the best but would be weird with static properties in classes. So, assume a good name here for now.

# So yeah...
What to do...

You tell me!

Poll is open, but comments for extra context are appreciated!

**Comment by nexxai** [comment]
It's awesome that you just posted this because I'm a relative newbie to Livewire and have been spending the last hour trying to figure out why my props weren't reactive, before landing on the doc page that explained that my assumptions of Livewire being more or less the "Laravel-version" of Vue were incorrect.

In a perfect world I would love for every prop to be reactive, but as you explained in point E that it's likely not happening, I think option C is the next best one.  

One of the things I love about LW is that I don't need constructors all over the place, and would prefer to avoid them if possible; denoting them with the two attributes removes ambiguity and makes it clear which ones are about to be Vue-like.

Thanks again for this amazing product and I look forward to the future where parent/child reactivity is commonplace! ❤️

**Reply by calebporzio** [reply]
Great feedback, thanks!

So with option C, does it bother you marking non-reactive props with `#[Prop]`? still? or it's fine

**Reply by nexxai** [reply]
Damn, I didn’t think about that part. I don’t think marking every prop with reactive or not would be a great experience; it feels very “un-Laravel” to not have the framework set intelligent defaults, leaving the developer to only mark where things diverge from the default. 

Does that make sense?

**Comment by lassemettovaara** [comment]
I'd advocate for option D, marking reactive properties explicitly makes much harder to do something stupid and also would love just using a constructor making this feel more like any old PHP class.

**Reply by calebporzio** [reply]
Ok, so with D, are you comfortable that the constructor is only called on the first request and not every subsequent one? Is that weird to you? or fine?

**Reply by adevade** [reply]
Instinctively option D feels the best personally. Mostly because it aligns with Blade components with the constructor. It makes LW components feel like regular Blade components on steroids! 

I haven't put down my vote yet though...

If it ends up being super hacky, it might not be worth it.

**Reply by calebporzio** [reply]
Ok cool cool - I don't think the implementation will be too bad actually

**Reply by lassemettovaara** [reply]
I guess I understand how someone might expect the constructor to be called on every subsequent request, but I wouldn't expect that at all. Even if it's not technically the same object, conceptually it is, so feels fine for me.

**Reply by markuskooche** [reply]
I also think that D is the best choice. Especially because you know constructors from everywhere.
I also think that experienced LW users understand that and why the method is called only once. People who don't know much about LW won't bother with subsequent.
An explanation in the documentation would also clarify all questions for beginners.

**Reply by thumbtech** [reply]
> why the method is called only once

To be clear, the method is called once **per request**, so not like mount, and I think very confusing and hacky, as Caleb notes. Or?

**Reply by jacksleight** [reply]
> Ok, so with D, are you comfortable that the constructor is only called on the first request and not every subsequent one? Is that weird to you? or fine?

I don’t think this is that weird, and have worked with other libraries that do exactly the same thing. eg. with Doctrine ORM `__construct` is only called when you create a brand new (to the database) entity, on fetch it uses reflection.

**Reply by calebporzio** [reply]
Oh that's really helpful insight (re: Doctrine) - great to know!

**Comment by Jubeki** [comment]
I personally think B is the best.

```php
class ShowPost extends Component
{
    #[Reactive]
    public $id;
}
```

But I think 'Reactive' is misleading (Take a look at [Filament Dependents Fields](https://filamentphp.com/docs/2.x/forms/advanced#dependant-fields--components))

I would say some along the line:
- Bound
- FromParent

On big question which I have is: What happens if the child component changes the value? Will the parent be updated, will only the child be updated or will an error be thrown? If only the child is updated, what happens if the parent changes?

I would also like it if an error is thrown, if a prop is passed to a livewire component which is not reactive. Something what Laravel does with mass assignment and fillable attributes [`Model::preventSilentlyDiscardingAttributes($this->app->isLocal())`](https://laravel.com/docs/10.x/eloquent#mass-assignment-exceptions)

**Reply by G3z** [reply]
Maybe `Inherited` is a better term to indicate that the prop comes from the parent component ?

**Reply by calebporzio** [reply]
Great question. If it's a non-reactive prop - no problem, you can still change the value.

If it's a reactive prop an error will be thrown: CannotChangeReactivePropertyException or something like that

**Reply by danharrin** [reply]
Came here to say this, I wouldn't use `Reactive` as we use that word to enable `wire:model` (default is `wire:model.defer`). I know Filament shouldn't dictate Livewire syntax, but a decent chunk of the community could end up confused, and I think there are better terms for this feature.

So I would also say `Inherit` or `Inherited` is the best here. Clearly defines the origin of the property as being passed in from a parent component.

**Reply by danharrin** [reply]
Now if you're thinking about `wire:model.defer` by default... `wire:model.reactive` would be good 😉

**Reply by cheesegrits** [reply]
I was initially really confused by Caleb's description because I was assuming "reactive" had the same meaning it does in Filament. As Dan says, Filament shouldn't dictate Livewire syntax, but there are a lot of people who's main exposure to LW is through Filament, and it will cause confusion. Not sure what a better word is, though.

**Reply by cheesegrits** [reply]
Oh, and C.

**Comment by peterfox** [comment]
Might hate me for this suggestion but...

Don't use annotations, use an array instead, with ['*'] as an option to use all:

```
class ShowPost extends Component
{
    public $id;

    protected $reactive = ['*'];
}
```

Then you can specify via the array which fields should be reactive.

```
class ShowPost extends Component
{
    public $id;

    protected $reactive = ['id'];
}
```

Annotations seem a good idea but I think in this instance it's adding a lot of boilerplate where an array can handle the job better, and handle both scenarios. It is also similar to how model guards work in Laravel.

**Reply by frknasir** [reply]
This seems like a good idea. hmm 🤔

**Reply by danharrin** [reply]
It's just my opinion, but I really don't like this array API. It decouples the behaviour of properties from their definitions. Using property attributes is a great approach to allow for more complex property configuration, such as co-locating the `$queryString` configuration with the property itself, or disabling property binding altogether.

**Reply by danharrin** [reply]
The reason why Eloquent uses this API in `$fillable` etc is because the properties aren't defined on the class. I'm sure that if they were, someone would have settled on an API using property attributes instead of arrays by now.

**Reply by peterfox** [reply]
That's a fair point @danharrin

**Reply by danharrin** [reply]
@frknasir you can't use configuration objects in property definitions though... PHP doesn't allow instantiating objects there

**Comment by thumbtech** [comment]
Yes A is not good, since "prop" already a thing and definitely not the same as this new thing ...
... and also "reactive" (B) not good, since I agree a lot of people already associate existing Livewire with "making page reactive", etc.
Wouldn't "aware" be the right term? isn't it intuitively similar to blade `@aware`, where therefore a Livewire [#Aware] attribute would be aware of parent property, which is already intuitively reactive, so my child id would update, etc. ... ?

**Comment by JunaidQadirB** [comment]


**Reply by calebporzio** [reply]
Mmm so how would this look in Livewire?

**Reply by ilyam2014** [reply]
![image](https://user-images.githubusercontent.com/89526742/226172835-93e0766f-58bf-40fe-b661-de8e14ac4dd0.png)

**Comment by pikseli** [comment]
For me the option D but with `[#Aware]` attribute feels like the most intuitive solution. It would use naming and concepts that are already familiar for Laravel devs.

**Reply by calebporzio** [reply]
Yeah, but "aware" isn't accurate in this case. @aware in a blade component is a way to just magically inherit scope of a certain property that was provided farther up the chain. Not a parent->child pass down if you know what I mean. It's Blade's answer to Vue and React's context API.

**Reply by thumbtech** [reply]
I'll take your word for it, but "way to just magically inherit scope of a certain property that was provided farther up the chain" and "a parent->child pass down" are pretty much synonymous to me.

**Comment by levu42** [comment]
I would love to have E) with a small change: all *public* properties are the reactive ones, and those that are not reactive, should be declared protected / private instead. However, this would break BC and I don't think it's worth it, so instead I think D) is the purest way to do it.

For me, Livewire always had the "you don't need to learn Vue / React, you can stick with what you know best, PHP, and create the same UX with it" aspect. So I think the target audience of Livewire is anyways people who know PHP quite well, and since PHP 8.0 with constructor property promotion, I think writing constructors is such a default path of declaring the interface of the class with the outside world, so I prefer the constructor solution.

As for the naming: It's only needed when you nest components, otherwise it doesn't really make sense, right? So why should the naming not reflect that fact and build upon that? I don't have a perfect name, but something related to "nested" or "parent" could be interesting. Maybe `#[Linked]`? Or, since it's Livewire, maybe just `#[Live]`?

**Comment by thijsdewitt** [comment]
Option E makes so much sense to me. With `id` being static and `:id` being reactive. Just like Alpine!
